staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,326 @@
+func (c *CELSchemaContext) ChildPropertyContext(propSchema *apiextensions.JSONSchemaProps, propertyName string) *CELSchemaContext {
+	if c == nil {
+		return nil
+	}
+	return c.childContext(propSchema, propertyTypeInfoAccessor{propertyName: propertyName})
+}
+
+// ChildAdditionalPropertiesContext constructs and returns a CELSchemaContext for the properties of an object if this CELSchemaContext is an object.
+// schema must be non-nil and have a non-nil schema.AdditionalProperties.
+func (c *CELSchemaContext) ChildAdditionalPropertiesContext(propsSchema *apiextensions.JSONSchemaProps) *CELSchemaContext {
+	if c == nil {
+		return nil
+	}
+	return c.childContext(propsSchema, additionalItemsTypeInfoAccessor{})
+}
+
+// ChildItemsContext constructs and returns a CELSchemaContext for the items of an array if this CELSchemaContext is an array.
+func (c *CELSchemaContext) ChildItemsContext(itemsSchema *apiextensions.JSONSchemaProps) *CELSchemaContext {
+	if c == nil {
+		return nil
+	}
+	return c.childContext(itemsSchema, itemsTypeInfoAccessor{})
+}
+
+// childContext constructs a new CELSchemaContext for the given child schema of the current schema context.
+// accessor optionally provides a way to access CELTypeInfo of the child from the current schema context's CELTypeInfo.
+// childContext returns a CELSchemaContext where the MaxCardinality is multiplied by the
+// factor that the schema increases the cardinality of its children. If the CELSchemaContext's
+// MaxCardinality is unbounded (nil) or the factor that the schema increase the cardinality
+// is unbounded, the resulting CELSchemaContext's MaxCardinality is also unbounded.
+func (c *CELSchemaContext) childContext(child *apiextensions.JSONSchemaProps, accessor typeInfoAccessor) *CELSchemaContext {
+	result := &CELSchemaContext{
+		parent:                    c,
+		typeInfoAccessor:          accessor,
+		withinValidationRuleScope: c.withinValidationRuleScope,
+		TotalCost:                 c.TotalCost,
+		MaxCardinality:            unbounded,
+		converter:                 c.converter,
+	}
+	if child != nil {
+		result.jsonSchema = child
+		if len(child.XValidations) > 0 {
+			result.withinValidationRuleScope = true
+		}
+	}
+	if c.jsonSchema == nil {
+		// nil schemas can be passed since we call ChildSchemaContext
+		// before ValidateCustomResourceDefinitionOpenAPISchema performs its nil check
+		return result
+	}
+	if c.MaxCardinality == unbounded {
+		return result
+	}
+	maxElements := extractMaxElements(c.jsonSchema)
+	if maxElements == unbounded {
+		return result
+	}
+	result.MaxCardinality = uint64ptr(multiplyWithOverflowGuard(*c.MaxCardinality, *maxElements))
+	return result
+}
+
+type typeInfoAccessor interface {
+	accessTypeInfo(parentTypeInfo *CELTypeInfo) *CELTypeInfo
+}
+
+type propertyTypeInfoAccessor struct {
+	// propertyName is the property name in the parent schema that this schema is declared at.
+	propertyName string
+}
+
```

Comment:
This is a general comment for these parent property accessors... I don't have a good sense for how paranoid they need to be about checking for nil values in parentTypeInfo or subfields. Can they assume parentTypeInfo is valid and has non-nil schema info / declType or do they need to check every possible nil subfield and return nil so callers can fall back the `converter` path for getting schema info?

Reply:
The rules are:

- CELTypeInfo must contain non-nil Schema and DeclType fields.
- accessTypeInfo must only be called with a non-nil parent CELTypeInfo

The code is enforcing this. I've updated the godoc to match and to clarify.